File: backend/app/routers/contacts.py
from typing import Any, List, Optional

from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile
from fastapi.responses import PlainTextResponse
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_session
from ..deps import get_current_user
from ..models import UserModel
from ..schemas.contact import (
    BulkIdsRequest,
    BulkMergeRequest,
    BulkUpdateRequest,
    ContactCreate,
    ContactResponse,
    ContactUpdate,
    DuplicateGroup,
    MergeContactsRequest,
    NoteCreate,
    NoteResponse,
    TimelineEvent,
)
from ..services.contact_service import ContactService

router = APIRouter()


@router.get("/", response_model=List[ContactResponse])
async def get_contacts(
    search: Optional[str] = Query(None),
    channel: Optional[str] = Query(None),
    subsection: Optional[str] = Query(None),
    contact_type: Optional[str] = Query(None),
    folder_id: Optional[int] = Query(None),
    is_favorite: Optional[bool] = Query(None),
    sort_by: str = Query("name"),
    sort_order: str = Query("asc"),
    skip: int = Query(0, ge=0),
    limit: Optional[int] = Query(None, ge=1),
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> list[ContactResponse]:
    service = ContactService(session, owner_id=int(current_user.id))
    return await service.get_all(
        search=search,
        channel=channel,
        subsection=subsection,
        contact_type=contact_type,
        folder_id=folder_id,
        is_favorite=is_favorite,
        sort_by=sort_by,
        sort_order=sort_order,
        skip=skip,
        limit=limit,
    )


@router.get("/export", response_class=PlainTextResponse)
async def export_contacts(
    ids: Optional[str] = Query(None),
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> str:
    service = ContactService(session, owner_id=int(current_user.id))
    id_list = [int(i) for i in ids.split(",")] if ids else None
    csv_data = await service.export_csv(ids=id_list)
    return csv_data


@router.get("/archive", response_model=List[ContactResponse])
async def get_archived_contacts(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=1000),
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> list[ContactResponse]:
    service = ContactService(session, owner_id=int(current_user.id))
    return await service.get_deleted(skip=skip, limit=limit)


@router.patch("/bulk-update")
async def bulk_update_contacts(
    data: BulkUpdateRequest,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> dict[str, Any]:
    service = ContactService(session, owner_id=int(current_user.id))
    fields: dict[str, object] = {}
    if data.is_known is not None:
        fields["is_known"] = data.is_known
    if data.folder_id is not None:
        fields["folder_id"] = data.folder_id
    if data.is_favorite is not None:
        fields["is_favorite"] = data.is_favorite
    if data.contact_type is not None:
        fields["contact_type"] = data.contact_type
    count = await service.bulk_update(data.ids, **fields)
    return {"updated": count}


@router.get("/duplicates", response_model=List[DuplicateGroup])
async def get_duplicates(
    contact_id: Optional[int] = Query(None),
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> list[dict[str, Any]]:
    service = ContactService(session, owner_id=int(current_user.id))
    return await service.get_duplicates(contact_id=contact_id)


# --- Параметризованные маршруты ниже литеральных ---

@router.get("/{contact_id}", response_model=ContactResponse)
async def get_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.get_by_id(contact_id)
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.post("/", response_model=ContactResponse, status_code=201)
async def create_contact(
    data: ContactCreate,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    try:
        return await service.create(data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.patch("/{contact_id}", response_model=ContactResponse)
async def update_contact(
    contact_id: int,
    data: ContactUpdate,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    try:
        contact = await service.update(contact_id, data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.patch("/{contact_id}/append", response_model=ContactResponse)
async def append_contact_fields(
    contact_id: int,
    data: ContactUpdate,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.append_fields(contact_id, data)
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.delete("/{contact_id}", status_code=204)
async def delete_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> None:
    service = ContactService(session, owner_id=int(current_user.id))
    deleted = await service.delete(contact_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Contact not found")


@router.delete("/{contact_id}/permanent", status_code=204)
async def permanent_delete_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> None:
    service = ContactService(session, owner_id=int(current_user.id))
    deleted = await service.permanent_delete(contact_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Contact not found")


@router.post("/{contact_id}/restore", response_model=ContactResponse)
async def restore_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.restore(contact_id)
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.post("/{contact_id}/block", response_model=ContactResponse)
async def block_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.update(contact_id, ContactUpdate(is_blocked=True))
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.post("/{contact_id}/unblock", response_model=ContactResponse)
async def unblock_contact(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.update(contact_id, ContactUpdate(is_blocked=False))
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


@router.post("/merge", response_model=ContactResponse)
async def merge_contacts(
    data: MergeContactsRequest,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    try:
        contact = await service.merge(data.primary_id, data.secondary_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    if not contact:
        raise HTTPException(status_code=400, detail="Failed to merge contacts")
    return contact


@router.post("/bulk-merge", response_model=ContactResponse)
async def bulk_merge_contacts(
    data: BulkMergeRequest,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> ContactResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    contact = await service.bulk_merge(data.primary_id, data.secondary_ids)
    if not contact:
        raise HTTPException(status_code=400, detail="Failed to merge contacts")
    return contact


@router.post("/bulk-delete")
async def bulk_delete_contacts(
    data: BulkIdsRequest,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> dict[str, Any]:
    service = ContactService(session, owner_id=int(current_user.id))
    count = await service.bulk_delete(data.ids)
    return {"deleted": count}


@router.post("/bulk-restore")
async def bulk_restore_contacts(
    data: BulkIdsRequest,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> dict[str, Any]:
    service = ContactService(session, owner_id=int(current_user.id))
    count = await service.bulk_restore(data.ids)
    return {"restored": count}


@router.get("/{contact_id}/timeline", response_model=List[TimelineEvent])
async def get_timeline(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> list[TimelineEvent]:
    service = ContactService(session, owner_id=int(current_user.id))
    return await service.get_timeline(contact_id)


@router.get("/{contact_id}/notes", response_model=List[NoteResponse])
async def get_notes(
    contact_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> list[NoteResponse]:
    service = ContactService(session, owner_id=int(current_user.id))
    return await service.get_notes(contact_id)


@router.post("/{contact_id}/notes", response_model=NoteResponse, status_code=201)
async def create_note(
    contact_id: int,
    data: NoteCreate,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> NoteResponse:
    service = ContactService(session, owner_id=int(current_user.id))
    note = await service.create_note(contact_id, data)
    if not note:
        raise HTTPException(status_code=404, detail="Contact not found")
    return note


@router.delete("/notes/{note_id}", status_code=204)
async def delete_note(
    note_id: int,
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> None:
    service = ContactService(session, owner_id=int(current_user.id))
    deleted = await service.delete_note(note_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Note not found")


@router.post("/import")
async def import_contacts(
    file: UploadFile = File(...),
    session: AsyncSession = Depends(get_session),
    current_user: UserModel = Depends(get_current_user),
) -> dict[str, int]:
    service = ContactService(session, owner_id=int(current_user.id))
    content_bytes = await file.read()
    content = content_bytes.decode("utf-8-sig")
    return await service.import_csv(content)

File: frontend/src/api/contacts.ts
import type { Contact, ContactCreate, ContactUpdate, ContactsQueryParams, ContactFolder, ContactFolderCreate, ContactFolderUpdate, BulkUpdateRequest, DuplicateGroup, TimelineEvent, ContactNote, ContactNoteCreate, ImportResult } from '../types/contact'
import { request } from './client'

export const contactApi = {
  getAll: (params?: ContactsQueryParams) => {
    const search = new URLSearchParams()
    if (params?.search) search.set('search', params.search)
    if (params?.channel) search.set('channel', params.channel)
    if (params?.subsection) search.set('subsection', params.subsection)
    if (params?.contact_type) search.set('contact_type', params.contact_type)
    if (params?.folder_id !== undefined && params.folder_id !== null) search.set('folder_id', String(params.folder_id))
    if (params?.is_favorite !== undefined) search.set('is_favorite', String(params.is_favorite))
    if (params?.sort_by) search.set('sort_by', params.sort_by)
    if (params?.sort_order) search.set('sort_order', params.sort_order)
    if (params?.skip !== undefined) search.set('skip', String(params.skip))
    if (params?.limit) search.set('limit', String(params.limit))
    const qs = search.toString()
    return request<Contact[]>(`/contacts${qs ? `?${qs}` : ''}`)
  },

  getById: (id: number) => request<Contact>(`/contacts/${id}`),

  create: (data: ContactCreate) =>
    request<Contact>('/contacts/', { method: 'POST', body: JSON.stringify(data) }),

  update: (id: number, data: ContactUpdate) =>
    request<Contact>(`/contacts/${id}`, { method: 'PATCH', body: JSON.stringify(data), _entityId: id }),

  delete: (id: number) =>
    request<void>(`/contacts/${id}`, { method: 'DELETE', _entityId: id }),

  search: (query: string, limit?: number) => {
    const qs = `search=${encodeURIComponent(query)}${limit ? `&limit=${limit}` : ''}`
    return request<Contact[]>(`/contacts?${qs}`)
  },

  merge: (primaryId: number, secondaryId: number) =>
    request<Contact>('/contacts/merge', {
      method: 'POST',
      body: JSON.stringify({ primary_id: primaryId, secondary_id: secondaryId }),
    }),

  getArchived: (params?: { skip?: number; limit?: number }) => {
    const search = new URLSearchParams()
    if (params?.skip) search.set('skip', String(params.skip))
    if (params?.limit) search.set('limit', String(params.limit))
    const qs = search.toString()
    return request<Contact[]>(`/contacts/archive${qs ? `?${qs}` : ''}`)
  },

  restore: (id: number) =>
    request<Contact>(`/contacts/${id}/restore`, { method: 'POST' }),

  block: (id: number) =>
    request<Contact>(`/contacts/${id}/block`, { method: 'POST', _entityId: id }),

  unblock: (id: number) =>
    request<Contact>(`/contacts/${id}/unblock`, { method: 'POST', _entityId: id }),

  bulkDelete: (ids: number[]) =>
    request<{ deleted: number }>('/contacts/bulk-delete', {
      method: 'POST',
      body: JSON.stringify({ ids }),
    }),

  bulkRestore: (ids: number[]) =>
    request<{ restored: number }>('/contacts/bulk-restore', {
      method: 'POST',
      body: JSON.stringify({ ids }),
    }),

  bulkMerge: (primaryId: number, secondaryIds: number[]) =>
    request<Contact>('/contacts/bulk-merge', {
      method: 'POST',
      body: JSON.stringify({ primary_id: primaryId, secondary_ids: secondaryIds }),
    }),

  getFolders: () =>
    request<ContactFolder[]>('/folders'),

  createFolder: (data: ContactFolderCreate) =>
    request<ContactFolder>('/folders/', { method: 'POST', body: JSON.stringify(data) }),

  updateFolder: (id: number, data: ContactFolderUpdate) =>
    request<ContactFolder>(`/folders/${id}`, { method: 'PATCH', body: JSON.stringify(data), _entityId: id }),

  deleteFolder: (id: number) =>
    request<void>(`/folders/${id}`, { method: 'DELETE', _entityId: id }),

  bulkUpdate: (data: BulkUpdateRequest) =>
    request<{ updated: number }>('/contacts/bulk-update', { method: 'PATCH', body: JSON.stringify(data) }),

  getDuplicates: (contactId?: number) => {
    const qs = contactId ? `?contact_id=${contactId}` : ''
    return request<DuplicateGroup[]>(`/contacts/duplicates${qs}`)
  },

  getTimeline: (contactId: number) =>
    request<TimelineEvent[]>(`/contacts/${contactId}/timeline`),

  getNotes: (contactId: number) =>
    request<ContactNote[]>(`/contacts/${contactId}/notes`),

  createNote: (contactId: number, data: ContactNoteCreate) =>
    request<ContactNote>(`/contacts/${contactId}/notes`, { method: 'POST', body: JSON.stringify(data) }),

  deleteNote: (noteId: number) =>
    request<void>(`/contacts/notes/${noteId}`, { method: 'DELETE', _entityId: noteId }),

  exportCsv: (ids?: number[]) => {
    const qs = ids?.length ? `?ids=${ids.join(',')}` : ''
    return request<string>(`/contacts/export${qs}`)
  },

  importCsv: (file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return request<ImportResult>('/contacts/import', { method: 'POST', body: formData })
  },
}
